File: 102-binary-tree-level-order-traversal/102-binary-tree-level-order-traversal.py
from collections import deque

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def levelOrder(self, root):
        if root is None:
            return []
        
        if root.left is None and root.right is None:
            return [[root.val]]
        
        answer = []
        q = deque([root])
        
        while len(q) != 0:
            currentLevel = []
            for i in range(len(q)):
                currentNode = q.popleft()
                currentLevel.append(currentNode.val)
                
                if currentNode.left is not None:
                    q.append(currentNode.left)
                if currentNode.right is not None:
                    q.append(currentNode.right)
            answer.append(currentLevel)
            
        
        return answer
        
        
        
        
        

File: 102-binary-tree-level-order-traversal/102-binary-tree-level-order-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function(root) {
    if (root == null) {
        return []    
    }
    
    const queue = [root]
    const levels = []
    while (queue.length > 0) {
        const qLength = queue.length 
        const level = []
        for (let i = 0; i < qLength; i++) {
            const popped = queue.shift()   
                if (popped) {
                level.push(popped.val)

                if (popped.left) {
                    queue.push(popped.left)
                } 

                if (popped.right) {
                    queue.push(popped.right)
                }
            }
        }
        levels.push(level)
    }
    return levels
};